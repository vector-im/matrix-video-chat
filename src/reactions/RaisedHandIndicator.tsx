/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only
Please see LICENSE in the repository root for full details.
*/

import {
  MouseEventHandler,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from "react";
import { DurationFormat } from "@formatjs/intl-durationformat";
import { useTranslation } from "react-i18next";

import { ReactionIndicator } from "./ReactionIndicator";

const durationFormatter = new DurationFormat(undefined, {
  minutesDisplay: "always",
  secondsDisplay: "always",
  hoursDisplay: "auto",
  style: "digital",
});

export function RaisedHandIndicator({
  raisedHandTime,
  miniature,
  showTimer,
  onClick,
}: {
  raisedHandTime?: Date;
  miniature?: boolean;
  showTimer?: boolean;
  onClick?: () => void;
}): ReactNode {
  const { t } = useTranslation();
  const [raisedHandDuration, setRaisedHandDuration] = useState("");

  const clickCallback = useCallback<MouseEventHandler<HTMLButtonElement>>(
    (event) => {
      if (!onClick) {
        return;
      }
      event.preventDefault();
      onClick();
    },
    [onClick],
  );

  // This effect creates a simple timer effect.
  useEffect(() => {
    if (!raisedHandTime || !showTimer) {
      return;
    }

    const calculateTime = (): void => {
      const totalSeconds = Math.ceil(
        (new Date().getTime() - raisedHandTime.getTime()) / 1000,
      );
      setRaisedHandDuration(
        durationFormatter.format({
          seconds: totalSeconds % 60,
          minutes: Math.floor(totalSeconds / 60),
        }),
      );
    };
    calculateTime();
    const to = setInterval(calculateTime, 1000);
    return (): void => clearInterval(to);
  }, [setRaisedHandDuration, raisedHandTime, showTimer]);

  if (!raisedHandTime) {
    return;
  }

  const content = (
    <ReactionIndicator emoji="âœ‹" miniature={miniature}>
      {showTimer && <p>{raisedHandDuration}</p>}
    </ReactionIndicator>
  );

  if (onClick) {
    return (
      <button
        aria-label={t("action.lower_hand")}
        style={{
          display: "contents",
          background: "none",
        }}
        onClick={clickCallback}
      >
        {content}
      </button>
    );
  }

  return content;
}
