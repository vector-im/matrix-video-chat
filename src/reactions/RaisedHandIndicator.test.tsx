/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only
Please see LICENSE in the repository root for full details.
*/

import { describe, expect, test } from "vitest";
import { render, configure } from "@testing-library/react";

import { RaisedHandIndicator } from "./RaisedHandIndicator";

configure({
  defaultHidden: true,
});

describe("RaisedHandIndicator", () => {
  test("renders nothing when no hand has been raised", () => {
    const { container } = render(<RaisedHandIndicator />);
    expect(container.firstChild).toBeNull();
  });
  test("renders an indicator when a hand has been raised", () => {
    const dateTime = new Date();
    const { container } = render(
      <RaisedHandIndicator raisedHandTime={dateTime} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test("renders an indicator when a hand has been raised with the expected time", () => {
    const dateTime = new Date(new Date().getTime() - 60000);
    const { container } = render(
      <RaisedHandIndicator raisedHandTime={dateTime} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test("renders a smaller indicator when minature is specified", () => {
    const dateTime = new Date();
    const { container } = render(
      <RaisedHandIndicator raisedHandTime={dateTime} minature />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
