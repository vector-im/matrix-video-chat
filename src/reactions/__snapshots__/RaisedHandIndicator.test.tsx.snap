// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RaisedHandIndicator > renders a smaller indicator when miniature is specified 1`] = `
<div
  class="reactionIndicatorWidget"
>
  <div
    class="reaction"
  >
    <span
      aria-label="common.reaction"
      role="img"
    >
      ✋
    </span>
  </div>
  <p>
    00:01
  </p>
</div>
`;

exports[`RaisedHandIndicator > renders an indicator when a hand has been raised 1`] = `
<div
  class="reactionIndicatorWidget reactionIndicatorWidgetLarge"
>
  <div
    class="reaction reactionLarge"
  >
    <span
      aria-label="common.reaction"
      role="img"
    >
      ✋
    </span>
  </div>
  <p>
    00:01
  </p>
</div>
`;

exports[`RaisedHandIndicator > renders an indicator when a hand has been raised with the expected time 1`] = `
<div
  class="reactionIndicatorWidget reactionIndicatorWidgetLarge"
>
  <div
    class="reaction reactionLarge"
  >
    <span
      aria-label="common.reaction"
      role="img"
    >
      ✋
    </span>
  </div>
  <p>
    01:01
  </p>
</div>
`;
